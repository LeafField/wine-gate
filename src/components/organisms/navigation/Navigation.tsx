"use client";
import React from "react";
import CategoryPanel from "../../molecules/category-panel/CategoryPanel";
import SearchPanel from "../../molecules/search-panel/SearchForm";
import UserPanel from "../../molecules/user-panel/UserPanel";
import { selectDummyData } from "../../../utils/dummyData";
import { categoryDummyData } from "../../../utils/dummyData";
import NavMenu from "../../molecules/nav-menu/NavMenu";
import { useStore } from "../../../store";

const Navigation = () => {
  const { menu } = useStore();
  return (
    <nav className="nav">
      <div
        className={`${menu === "close" ? "invisible" : "show-panel"} z-nav tablet:visible tablet:static tablet:space-y-6`}
        id="navigation"
      >
        <UserPanel />
        <CategoryPanel categories={categoryDummyData} />
        <SearchPanel selectData={selectDummyData} />
      </div>
      <NavMenu />
    </nav>
  );
};

export default Navigation;
