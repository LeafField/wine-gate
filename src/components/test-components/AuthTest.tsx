"use client";
import React, { useEffect } from "react";
import { supabase } from "../../utils/supabase";

const AuthTest = () => {
  useEffect(() => {
    const getSession = async () => {
      const session = await supabase.auth.getSession();
      console.log(session.data.session?.user.user_metadata);
    };
    getSession();
  }, []);

  const clickHandler = async () => {
    await supabase.auth.signOut();
  };

  return <button onClick={clickHandler}>ログアウト</button>;
};

export default AuthTest;
