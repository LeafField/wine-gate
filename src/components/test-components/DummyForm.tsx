import React, { FC, PropsWithChildren, FormEvent, useState } from "react";

type FormState = {
  [key: string]: FormDataEntryValue;
};

const DummyForm: FC<PropsWithChildren> = ({ children }) => {
  const [input, setInput] = useState<string>("");
  const submitHandler = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const formData = new FormData(e.currentTarget);
    // const data = Object.fromEntries(formData.entries());
    const data = formData.get("search");
    if (data) {
      setInput(data.toString());
    }
  };

  return (
    <form onSubmit={submitHandler}>
      {children}
      <p>{input}</p>
    </form>
  );
};

export default DummyForm;
