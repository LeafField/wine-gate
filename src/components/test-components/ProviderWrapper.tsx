import { QueryClientProvider } from "@tanstack/react-query";
import { FC, PropsWithChildren } from "react";
import { queryClient } from "../../utils/queryClient";
import { MantineProvider } from "@mantine/core";

export const ProviderWrapper: FC<PropsWithChildren> = ({ children }) => {
  return (
    <QueryClientProvider client={queryClient}>
      <MantineProvider>{children}</MantineProvider>
    </QueryClientProvider>
  );
};
