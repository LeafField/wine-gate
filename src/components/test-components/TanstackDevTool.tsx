import React from "react";
import { ReactQueryDevtools } from "@tanstack/react-query-devtools";

const TanstackDevTool = () => {
  return <ReactQueryDevtools initialIsOpen={false} />;
};

export default TanstackDevTool;
