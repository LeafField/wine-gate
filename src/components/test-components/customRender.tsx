import { render as testingLibraryRender } from "@testing-library/react";
import { MantineProvider } from "@mantine/core";

export function CustomRender(ui: React.ReactNode) {
  return testingLibraryRender(<>{ui}</>, {
    wrapper: ({ children }: { children: React.ReactNode }) => (
      <MantineProvider defaultColorScheme="light">{children}</MantineProvider>
    ),
  });
}
