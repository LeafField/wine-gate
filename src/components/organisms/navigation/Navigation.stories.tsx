import { Meta, StoryObj } from "@storybook/react";
import Navigation from "./Navigation";
import { spyOn, within } from "@storybook/test";
import { supabase } from "../../../utils/supabase";

const meta = {
  title: "organisms/Navigation",
  component: Navigation,
  parameters: {
    nextjs: {
      appDirectory: true,
    },
  },
} as Meta<typeof Navigation>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  play: async ({ canvasElement }) => {
    within(canvasElement);
    spyOn(supabase.auth, "signOut").mockImplementationOnce(() => {
      console.log("signOut");
      return Promise.resolve({ error: null, data: null });
    });
  },
};
