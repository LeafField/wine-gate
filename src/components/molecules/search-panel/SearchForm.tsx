"use client";
import React, { FC, FormEvent, useState, useRef, useEffect } from "react";
import { useRouter } from "next/navigation";
import SearchButton from "../../atoms/search-button/SearchButton";
import SearchInput from "../../atoms/search-input/SearchInput";
import { ComboboxData } from "@mantine/core";
import { searchSchema } from "../../../utils/schema";
import { useStore } from "../../../store";
import { useQueryClient } from "@tanstack/react-query";

type Props = {
  selectData: ComboboxData;
};

const SearchPanel: FC<Props> = ({ selectData }) => {
  const [errorState, setErrorState] = useState<string | null>(null);
  const inputRef = useRef<HTMLInputElement>(null);
  const router = useRouter();
  const { menu, setMenu, clearSearchValue } = useStore();
  const queryClient = useQueryClient();

  const submitHandler = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    inputRef.current?.blur();
    queryClient.removeQueries({ queryKey: ["search"] });

    const formData = new FormData(event.currentTarget);
    const data = Object.fromEntries(formData.entries());
    const schema = searchSchema.safeParse(data);
    if (schema.success) {
      const slug = schema.data;
      router.push(`/search?search=${slug.search}`);
      setErrorState(null);
      setMenu("close");
    } else {
      setErrorState(schema.error.errors[0].message);
    }
    clearSearchValue();
  };

  useEffect(() => {
    router.prefetch("/search");
  }, [router]);

  return (
    <div
      className={`nav-panel max-tablet:animate-fadeIn ${menu === "free" ? "max-tablet:flex" : "max-tablet:hidden"}`}
      id="free-search-panel"
    >
      <h2 className="nav-title">フリーワード検索</h2>
      <form
        className="flex w-full flex-col items-end justify-end gap-y-3"
        onSubmit={submitHandler}
        role="search"
      >
        <SearchInput ref={inputRef} error={errorState} />
        <SearchButton />
      </form>
    </div>
  );
};

export default SearchPanel;
