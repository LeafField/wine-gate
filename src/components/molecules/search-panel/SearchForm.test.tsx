import { render, screen, waitFor } from "@testing-library/react";
import SearchForm from "./SearchForm";
import { selectDummyData } from "../../../utils/dummyData";
import * as router from "next/navigation";
import userEvent from "@testing-library/user-event";
import { ProviderWrapper } from "../../test-components/ProviderWrapper";

jest.mock("next/navigation");
const mockRouterPush = jest.fn();

describe("SearchFormの結合テスト", () => {
  beforeEach(() => {
    jest.spyOn(router, "useRouter").mockImplementation(() => ({
      push: mockRouterPush,
      back: jest.fn(),
      prefetch: jest.fn(),
      reload: jest.fn(),
      replace: jest.fn(),
      pathname: "/",
      route: "/",
      query: {},
      asPath: "/",
      basePath: "",
      forward: jest.fn(),
      refresh: jest.fn(),
    }));

    render(<SearchForm selectData={selectDummyData} />, {
      wrapper: ProviderWrapper,
    });
  });
  afterEach(() => {
    jest.clearAllMocks();
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  test("コンポーネントがクラッシュせずにレンダリングされる", () => {
    const formElement = screen.getByRole("search");
    expect(formElement).toBeInTheDocument();
  });

  test("正常系:submitActionが正しく動作しrouter.pushが実行される", async () => {
    const textBox = screen.getByRole("textbox", { name: "search" });
    const submitButton = screen.getByRole("button", { name: "検索" });
    await userEvent.type(textBox, "コノスル");
    await userEvent.click(submitButton);

    await waitFor(() => {
      expect(mockRouterPush).toHaveBeenCalledWith(
        "/search?sort=new&category=&search=コノスル",
      );
    });
  });

  test("異常系:inputタグの入力が4文字未満の時、バリデーションエラーを表示する", async () => {
    const textBox = screen.getByRole("textbox", { name: "search" });
    const submitButton = screen.getByRole("button", { name: "検索" });
    await userEvent.type(textBox, "コノス");
    await userEvent.click(submitButton);

    await waitFor(() => {
      expect(screen.getByText("4文字以上入力してください")).toBeInTheDocument();
    });
  });

  test("スナップショットテスト", () => {
    const { container } = render(<SearchForm selectData={selectDummyData} />, {
      wrapper: ProviderWrapper,
    });
    expect(container).toMatchSnapshot();
  });
});
