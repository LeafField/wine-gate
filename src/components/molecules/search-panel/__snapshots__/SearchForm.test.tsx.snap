// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchFormの結合テスト スナップショットテスト 1`] = `
<div>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
  </style>
  <div
    class="nav-panel max-tablet:animate-fadeIn max-tablet:hidden"
    id="free-search-panel"
  >
    <h2
      class="nav-title"
    >
      フリーワード検索
    </h2>
    <form
      class="flex w-full flex-col items-end justify-end gap-y-3"
      role="search"
    >
      <div
        class="w-full focus:border-blue m_46b77525 mantine-InputWrapper-root mantine-TextInput-root"
      >
        <div
          class="m_6c018570 mantine-Input-wrapper mantine-TextInput-wrapper"
          data-variant="default"
          data-with-left-section="true"
          style="--input-right-section-pointer-events: all;"
        >
          <div
            class="m_82577fc2 mantine-Input-section mantine-TextInput-section"
            data-position="left"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-magnifying-glass "
              data-icon="magnifying-glass"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M416 208c0 45.9-14.9 88.3-40 122.7L502.6 457.4c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L330.7 376c-34.4 25.2-76.8 40-122.7 40C93.1 416 0 322.9 0 208S93.1 0 208 0S416 93.1 416 208zM208 352a144 144 0 1 0 0-288 144 144 0 1 0 0 288z"
                fill="currentColor"
              />
            </svg>
          </div>
          <input
            aria-invalid="false"
            aria-label="search"
            class="m_8fb7ebe7 mantine-Input-input mantine-TextInput-input"
            data-variant="default"
            id="search"
            name="search"
            placeholder="検索"
            value=""
          />
        </div>
      </div>
      <button
        class="border-2 border-gray px-3 py-1 transition-colors duration-300 hover:border-blue focus:border-blue focus:outline-none"
        type="submit"
      >
        検索
      </button>
    </form>
  </div>
</div>
`;
